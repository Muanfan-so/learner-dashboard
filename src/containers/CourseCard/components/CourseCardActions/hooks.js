import { Locked } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import { hooks as appHooks } from 'data/redux';
import useSelectSession from 'containers/SelectSession/hooks';
import messages from './messages';

export const useCardActionData = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const {
    canUpgrade,
    isAudit,
    isAuditAccessExpired,
    isVerified,
  } = appHooks.useCardEnrollmentData(courseNumber);
  const { isPending, isArchived } = appHooks.useCardCourseRunData(courseNumber);
  const {
    isEntitlement,
    canViewCourse,
    isFulfilled,
    isExpired,
    canChange,
    hasSessions,
  } = appHooks.useCardEntitlementsData(courseNumber);
  const { openSessionModal } = useSelectSession({ courseNumber });

  let primary;
  let secondary = null;
  if (isEntitlement) {
    if (!isFulfilled) {
      primary = {
        children: formatMessage(messages.selectSession),
        disabled: !(canChange && hasSessions),
        onClick: openSessionModal,
      };
    } else {
      primary = {
        children: formatMessage(messages.viewCourse),
        disabled: !canViewCourse || isExpired,
      };
    }
  } else {
    if (!isVerified) {
      secondary = {
        iconBefore: Locked,
        variant: 'outline-primary',
        disabled: !canUpgrade,
        children: formatMessage(messages.upgrade),
      };
    }
    if (isPending) {
      primary = { children: formatMessage(messages.beginCourse) };
    } else if (!isArchived) {
      primary = {
        children: formatMessage(messages.resume),
        disabled: isAudit && isAuditAccessExpired,
      };
    } else {
      primary = { children: formatMessage(messages.viewCourse) };
    }
  }
  return { primary, secondary };
};

export default useCardActionData;
