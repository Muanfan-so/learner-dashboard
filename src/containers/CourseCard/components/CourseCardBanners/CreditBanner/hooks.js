import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = appHooks.useCardCreditData(cardId);
  const { supportEmail } = appHooks.usePlatformSettingsData();
  if (!credit.isEligible) { return null; }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
