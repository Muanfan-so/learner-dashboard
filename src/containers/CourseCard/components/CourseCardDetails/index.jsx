import React from 'react';
import PropTypes from 'prop-types';

import { Button } from '@edx/paragon';

import useCardDetailsData from './hooks';

import messages from './messages';

export const CourseCardDetails = ({ courseNumber }) => {
  const {
    providerName,
    accessMessage,
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    formatMessage,
  } = useCardDetailsData({ courseNumber });

  return (
    <span data-testid="CourseCardDetails">
      {providerName} • {courseNumber} • {accessMessage}
      {isEntitlement && isFulfilled && canChange ? (
        <>
          {' • '}
          <Button variant="link" size="inline" className="m-0 p-0" onClick={openSessionModal}>
            {formatMessage(messages.changeOrLeaveSessionButton)}
          </Button>
        </>
      ) : null}
    </span>
  );
};

CourseCardDetails.propTypes = {
  courseNumber: PropTypes.string.isRequired,
};

CourseCardDetails.defaultProps = {};

export default CourseCardDetails;
