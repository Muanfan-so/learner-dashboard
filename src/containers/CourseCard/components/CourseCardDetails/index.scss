@use "~@edx/brand/paragon/variables";
@use "~@openedx/paragon/scss/core/core";
@use "~@edx/brand/paragon/overrides";

a.course-card-title {
  color: $black;
}
