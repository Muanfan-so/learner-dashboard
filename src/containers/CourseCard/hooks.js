import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

export const useCardData = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const { title, bannerUrl } = appHooks.useCardCourseData(courseNumber);

  return {
    title,
    bannerUrl,
    formatMessage,
  };
};

export default useCardData;
