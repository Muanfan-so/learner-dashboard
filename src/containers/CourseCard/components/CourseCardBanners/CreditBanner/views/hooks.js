import React from 'react';

import { StrictDict } from 'utils';
import { apiHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  creditRequestData: (val) => React.useState(val), // eslint-disable-line
});

export const useCreditRequestData = (cardId) => {
  const [requestData, setRequestData] = module.state.creditRequestData(null);
  const createCreditApiRequest = apiHooks.useCreateCreditRequest(cardId);
  const createCreditRequest = (e) => {
    e.preventDefault();
    createCreditApiRequest().then(setRequestData);
  };
  return { requestData, createCreditRequest };
};

export default {
  useCreditRequestData,
};
