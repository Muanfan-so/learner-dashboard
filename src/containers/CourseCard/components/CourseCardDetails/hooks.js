import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';
import useSelectSession from 'containers/SelectSession/hooks';

import * as module from './hooks';
import messages from './messages';

export const useAccessMessage = ({ courseNumber }) => {
  const { formatMessage, formatDate } = useIntl();
  const {
    accessExpirationDate,
    isAudit,
    isAuditAccessExpired,
  } = appHooks.useCardEnrollmentData(courseNumber);
  const { isArchived, endDate } = appHooks.useCardCourseRunData(courseNumber);

  if (isAudit) {
    return formatMessage(
      isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
      { accessExpirationDate: formatDate(accessExpirationDate) },
    );
  }

  return formatMessage(
    isArchived ? messages.courseEnded : messages.courseEnds,
    { endDate: formatDate(endDate) },
  );
};

export const useCardDetailsData = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(courseNumber).name;
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementsData(courseNumber);

  const { openSessionModalWithLeaveOption: openSessionModal } = useSelectSession({ courseNumber });

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: module.useAccessMessage({ courseNumber }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    formatMessage,
  };
};

export default useCardDetailsData;
