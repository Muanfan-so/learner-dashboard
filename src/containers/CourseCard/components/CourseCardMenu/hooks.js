import React from 'react';
import { StrictDict } from 'utils';

import track from 'tracking';
import { reduxHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val), // eslint-disable-line
  isEmailSettingsVisible: (val) => React.useState(val), // eslint-disable-line
});

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useHandleToggleDropdown = (cardId) => {
  const eventName = track.course.courseOptionsDropdownClicked;
  const trackCourseEvent = reduxHooks.useTrackCourseEvent(eventName, cardId);
  return (isOpen) => {
    if (isOpen) { trackCourseEvent(); }
  };
};

export const useCourseCardMenu = (cardId) => {
  const { courseName } = reduxHooks.useCardCourseData(cardId);
  const { isEnrolled, isEmailEnabled } = reduxHooks.useCardEnrollmentData(cardId);
  const { twitter, facebook } = reduxHooks.useCardSocialSettingsData(cardId);
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const { isEarned } = reduxHooks.useCardCertificateData(cardId);
  const handleTwitterShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'twitter',
  );
  const handleFacebookShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'facebook',
  );

  const showUnenrollItem = isEnrolled && !isEarned;
  const showDropdown = showUnenrollItem || isEmailEnabled || facebook.isEnabled || twitter.isEnabled;

  return {
    courseName,
    isMasquerading,
    isEmailEnabled,
    showUnenrollItem,
    showDropdown,
    facebook,
    twitter,
    handleTwitterShare,
    handleFacebookShare,
  };
};
