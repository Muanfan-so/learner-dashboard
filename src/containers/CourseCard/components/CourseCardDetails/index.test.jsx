import React from 'react';
import { shallow } from 'enzyme';

import CourseCardDetails from '.';

import hooks from './hooks';

jest.mock('./hooks', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const courseNumber = 'test-course-number';

describe('CourseCard Details component', () => {
  it('has change session button on entitlement course', () => {
    const mockHook = (args) => () => ({
      providerName: 'provider-name',
      accessMessage: 'access-message',
      openSessionModal: jest.fn().mockName('useSelectSession.openSessionModal'),
      formatMessage: (message, values) => <div {...{ message, values }} />,
      isEntitlement: true,
      isFulfilled: true,
      canChange: true,
      ...args,
    });
    hooks.mockImplementationOnce(mockHook({ isEntitlement: true }));
    const el = shallow(<CourseCardDetails courseNumber={courseNumber} />);
    expect(el).toMatchSnapshot();
    // it has 3 separator, 4 column
    expect(el.text().match(/•/g)).toHaveLength(3);
  });

  it('does not have change session button on regular course', () => {
    const mockHook = (args) => () => ({
      providerName: 'provider-name',
      accessMessage: 'acess-message',
      openSessionModal: jest.fn().mockName('useSelectSession.openSessionModal'),
      formatMessage: (message, values) => <div {...{ message, values }} />,
      isEntitlement: true,
      isFulfilled: true,
      canChange: true,
      ...args,
    });
    hooks.mockImplementationOnce(mockHook({ isEntitlement: false }));
    const el = shallow(<CourseCardDetails courseNumber={courseNumber} />);
    expect(el).toMatchSnapshot();
    // it has 2 separator, 3 column
    expect(el.text().match(/•/g)).toHaveLength(2);
  });
});
