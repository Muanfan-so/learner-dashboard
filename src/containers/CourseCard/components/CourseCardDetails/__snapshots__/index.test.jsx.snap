// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard Details component does not have change session button on regular course 1`] = `
<span
  data-testid="CourseCardDetails"
>
  provider-name
   • 
  test-course-number
   • 
  acess-message
</span>
`;

exports[`CourseCard Details component has change session button on entitlement course 1`] = `
<span
  data-testid="CourseCardDetails"
>
  provider-name
   • 
  test-course-number
   • 
  access-message
   • 
  <Button
    className="m-0 p-0"
    onClick={[MockFunction useSelectSession.openSessionModal]}
    size="inline"
    variant="link"
  >
    <div
      message={
        Object {
          "defaultMessage": "Change or leave session",
          "description": "Button for trigger change or leave session for entitlement course",
          "id": "learner-dash.courseCard.CourseCardDetails.changeOrLeaveSessionButton",
        }
      }
    />
  </Button>
</span>
`;
